Lifter + PageBuster

zTehRyaN opened the pull request:
Some additions to the Lifter logic to handle PageBuster custom ELF file.

zTehRyaN committed:
BinaryFile: PageBuster initial integration

Added a static local epoch counter and a draft of integration
in the lifter

aleclearmind commented on tools/revng-lift/BinaryFile.cpp:
```diff
@@ -39,6 +39,12 @@ using LabelList = BinaryFile::LabelList;
+// TODO find a finer way
+constexpr int PT_PAGEBUSTER { 0x8 };
+
+static int g_epoch_counter { 0 };
```
Do not use global variables, this can be a local variable just outside the loop.
Also make sure you follow the [LLVM Coding Conventions](https://llvm.org/docs/CodingStandards.html).

```suggestion
static int EpocCounter { 0 };
```

aleclearmind commented on tools/revng-lift/BinaryFile.cpp:
```diff
@@ -957,10 +963,13 @@ void BinaryFile::parseELF(object::ObjectFile *TheBinary,
     switch (ProgramHeader.p_type) {
+    case PT_PAGEBUSTER:
     case ELF::PT_LOAD: {
+      bool pagebuster { ProgramHeader.p_type == PT_PAGEBUSTER };
```
Just use regular assignments unless there's a reason to.
```suggestion
      bool IsPageBuster = (ProgramHeader.p_type == PT_PAGEBUSTER);
```

aleclearmind commented on tools/revng-lift/BinaryFile.cpp:
```diff
@@ -1002,7 +1011,13 @@ void BinaryFile::parseELF(object::ObjectFile *TheBinary,
+      }
+
     } break;
+    [[fallthrough]];
```
Why? It's not falling through, there's a `break`.

aleclearmind commented on tools/revng-lift/BinaryFile.cpp:
```diff
@@ -1002,7 +1011,13 @@ void BinaryFile::parseELF(object::ObjectFile *TheBinary,
         ProgramHeaders.Address = PhdrAddress;
       }
+
+      if (pagebuster) {
+        g_epoch_counter++;
+      }
```
Maybe we should outline the whole body of this `case` into a function and call it differently depending on whether we're in `case PT_PAGEBUSTER` or `case PT_LOAD`.

Commit:
MetaAddress.cpp: Fixed fields processing order

Commit:
OriginalInstructionAddresses repopulate function

Commit:
Manual entry points addition

Commit:
SwitchManager constructor fixup

Commit:
Proper Segment mapping + entrypoints optionality

We can consider an epoch to be unique. So we can search for the relati-
ve segment by passing only the epoch to the method. Right now, we will
fine the unique corrisponding segment, but in a possible future with
many epochs within a single segment, this will also works.

Commit:
Reducing required `mmap` update operations

Commit:
Added epoch info on global Segment Variables

Commit:
CMake: AddressSanitizer added

Commit:
CodeGenerator: Custom mapping
